File: MiniDB_Projects/Team_LockAndRoll/benchmarks/bench.cpp
// concurrency benchmark: 2PL vs MVCC.
// workload is long read-only scans next to point writers — where MVCC wins
// because its readers don't block on the writers' row locks. fsync-on-commit is
// off so we measure concurrency-control cost, not disk latency.
#include <atomic>
#include <chrono>
#include <iostream>
#include <random>
#include <thread>
#include <vector>

#include "engine.h"

using namespace minidb;
using Clock = std::chrono::steady_clock;

namespace {

struct Result {
  double seconds = 0;
  long reads = 0;
  long writes = 0;
  long aborts = 0;
};

Result run_workload(CCMode mode, const std::string& dir, int rows, int readers,
                    int writers, int reads_each, int writes_each) {
  Database db(dir, mode, 1024);
  db.set_durable(false);  // isolate CC cost from fsync latency
  db.execute("CREATE TABLE t (id INTEGER PRIMARY KEY, v INTEGER)");
  {
    std::string ins = "INSERT INTO t VALUES ";
    for (int i = 0; i < rows; i++)
      ins += (i ? "," : "") + std::string("(") + std::to_string(i) + ",0)";
    db.execute(ins);
  }
  TableInfo* t = db.catalog().get_table("t");
  std::atomic<long> aborts{0};

  auto reader = [&]() {
    for (int i = 0; i < reads_each; i++) {
      Transaction* txn = db.begin(true);
      long n = 0;
      db.scan_table(txn, t, [&](RID, const Tuple&) { n++; return true; });
      db.commit(txn);
      (void)n;
    }
  };
  auto writer = [&](int seed) {
    std::mt19937 rng(seed);
    std::uniform_int_distribution<int> key_dist(0, rows - 1);
    long local_aborts = 0;
    for (int i = 0; i < writes_each; i++) {
      for (;;) {
        Transaction* txn = db.begin(true);
        try {
          int k = key_dist(rng);
          db.delete_row(txn, t, k);
          db.insert_row(txn, t, Tuple({Value((int64_t)k), Value((int64_t)i)}));
          db.commit(txn);
          break;
        } catch (const AbortException&) {
          db.abort(txn);
          local_aborts++;
        }
      }
    }
    aborts += local_aborts;
  };

  auto start = Clock::now();
  std::vector<std::thread> pool;
  for (int i = 0; i < readers; i++) pool.emplace_back(reader);
  for (int i = 0; i < writers; i++) pool.emplace_back(writer, i + 100);
  for (auto& th : pool) th.join();
  double secs = std::chrono::duration<double>(Clock::now() - start).count();
  return {secs, (long)readers * reads_each, (long)writers * writes_each, aborts.load()};
}

void report(const char* name, const Result& r) {
  double read_tps = r.reads / r.seconds;
  double write_tps = r.writes / r.seconds;
  std::cout << "  " << name << ": " << r.seconds << "s | reads " << r.reads << " ("
            << (long)read_tps << "/s) | writes " << r.writes << " (" << (long)write_tps
            << "/s) | aborts " << r.aborts << "\n";
}

}  // namespace

int main(int argc, char** argv) {
  int rows = 500, readers = 6, writers = 2, reads_each = 400, writes_each = 4000;
  for (int i = 1; i < argc; i++) {
    std::string a = argv[i];
    auto val = [&]() { return std::stoi(argv[++i]); };
    if (a == "--rows") rows = val();
    else if (a == "--readers") readers = val();
    else if (a == "--writers") writers = val();
    else if (a == "--reads-each") reads_each = val();
    else if (a == "--writes-each") writes_each = val();
  }

  std::cout << "MiniDB concurrency benchmark (long readers + point writers)\n"
            << "  rows=" << rows << " readers=" << readers << " writers=" << writers
            << " reads/reader=" << reads_each << " writes/writer=" << writes_each << "\n\n";

  Result tpl = run_workload(CCMode::TWO_PL, "bench_2pl", rows, readers, writers, reads_each,
                            writes_each);
  Result mvcc = run_workload(CCMode::MVCC, "bench_mvcc", rows, readers, writers, reads_each,
                             writes_each);
  std::cout << "Results:\n";
  report("2PL ", tpl);
  report("MVCC", mvcc);
  double total_2pl = (tpl.reads + tpl.writes) / tpl.seconds;
  double total_mvcc = (mvcc.reads + mvcc.writes) / mvcc.seconds;
  std::cout << "\n  Total throughput  2PL: " << (long)total_2pl << " txn/s,  MVCC: "
            << (long)total_mvcc << " txn/s\n";
  std::cout << "  MVCC speedup over 2PL: " << (total_mvcc / total_2pl) << "x\n";
  return 0;
}

File: MiniDB_Projects/Team_SuperSymmetry/benchmarks/bench.py
"""
MiniDB benchmark suite.

Three experiments, each producing real measured numbers:

  1. Concurrency: read throughput and blocking under contention,
     strict 2PL vs MVCC (the Extension Track B comparison).
  2. Access methods: latency of point lookups served by a B+ tree
     IndexScan vs a full SeqScan.
  3. Buffer pool: hit rate as the working set grows relative to the pool.

Results are printed and also written to benchmarks/results.md.

Run:
    python benchmarks/bench.py
"""
import os
import random
import shutil
import statistics
import sys
import threading
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

from minidb import Database, TransactionError

OUT = os.path.join(os.path.dirname(__file__), "results.md")
random.seed(1234)


def fresh(directory, isolation):
    shutil.rmtree(directory, ignore_errors=True)
    return Database(directory, isolation=isolation)


# ---------------------------------------------------------------------------
# Experiment 1 — concurrency: 2PL vs MVCC
# ---------------------------------------------------------------------------
def load_accounts(db, n):
    db.execute("CREATE TABLE acct (id INT PRIMARY KEY, bal INT)")
    db.execute("INSERT INTO acct VALUES " +
               ",".join(f"({i},{1000})" for i in range(n)))
    db.analyze()


def run_writers_only(isolation, n_rows=3000, hot=40, writers=2,
                     duration=2.0, write_hold=0.012):
    """Write throughput measured without busy readers, so CPython's GIL (under
    which MVCC's non-blocking readers would otherwise starve writer threads)
    does not distort the comparison."""
    directory = f"/tmp/minidb_benchw_{isolation.lower()}"
    db = fresh(directory, isolation)
    load_accounts(db, n_rows)
    stop = threading.Event()
    writes_ok = [0]
    lock = threading.Lock()

    def writer():
        local = 0
        while not stop.is_set():
            rid = random.randrange(hot)
            try:
                txn = db.begin()
                db.execute(f"DELETE FROM acct WHERE id = {rid}", txn=txn)
                db.execute(f"INSERT INTO acct VALUES ({rid}, {random.randint(0,2000)})",
                           txn=txn)
                time.sleep(write_hold)
                db.commit(txn)
                local += 1
            except TransactionError:
                pass
        with lock:
            writes_ok[0] += local

    threads = [threading.Thread(target=writer) for _ in range(writers)]
    t0 = time.perf_counter()
    for t in threads:
        t.start()
    time.sleep(duration)
    stop.set()
    for t in threads:
        t.join()
    elapsed = time.perf_counter() - t0
    db.close()
    return {"isolation": isolation, "write_tps": writes_ok[0] / elapsed,
            "writes_ok": writes_ok[0]}


def run_contention(isolation, n_rows=3000, hot=40, readers=8, writers=2,
                   duration=2.5, write_hold=0.012):
    directory = f"/tmp/minidb_bench_{isolation.lower()}"
    db = fresh(directory, isolation)
    load_accounts(db, n_rows)

    stop = threading.Event()
    reads_ok = [0]
    reads_latency_s = [0.0]
    writes_ok = [0]
    lock = threading.Lock()

    def reader():
        local = 0
        latency = 0.0
        while not stop.is_set():
            rid = random.randrange(hot)        # contend on the hot set
            t0 = time.perf_counter()
            try:
                txn = db.begin()
                _ = list(db.execute(f"SELECT bal FROM acct WHERE id = {rid}", txn=txn))
                db.commit(txn)
                local += 1
                latency += time.perf_counter() - t0
            except TransactionError:
                pass  # aborted under contention; not counted as completed
        with lock:
            reads_ok[0] += local
            reads_latency_s[0] += latency

    def writer():
        local = 0
        while not stop.is_set():
            rid = random.randrange(hot)
            try:
                txn = db.begin()
                db.execute(f"DELETE FROM acct WHERE id = {rid}", txn=txn)
                db.execute(f"INSERT INTO acct VALUES ({rid}, {random.randint(0,2000)})",
                           txn=txn)
                time.sleep(write_hold)         # hold the row briefly
                db.commit(txn)
                local += 1
            except TransactionError:
                pass
        with lock:
            writes_ok[0] += local

    threads = ([threading.Thread(target=reader) for _ in range(readers)] +
               [threading.Thread(target=writer) for _ in range(writers)])
    t0 = time.perf_counter()
    for t in threads:
        t.start()
    time.sleep(duration)
    stop.set()
    for t in threads:
        t.join()
    elapsed = time.perf_counter() - t0
    db.close()

    return {
        "isolation": isolation,
        "elapsed_s": elapsed,
        "reads_ok": reads_ok[0],
        "read_tps": reads_ok[0] / elapsed,
        "avg_read_latency_ms": (reads_latency_s[0] / reads_ok[0] * 1000) if reads_ok[0] else float("nan"),
        "writes_ok": writes_ok[0],
        "write_tps": writes_ok[0] / elapsed,
    }


# ---------------------------------------------------------------------------
# Experiment 2 — index scan vs seq scan
# ---------------------------------------------------------------------------
def run_access_methods(n_rows=4000, probes=800):
    db = fresh("/tmp/minidb_bench_idx", "2PL")
    db.execute("CREATE TABLE t (id INT PRIMARY KEY, shadow INT, v INT)")
    # shadow holds the same value as id but is NOT indexed -> forces SeqScan
    db.execute("INSERT INTO t VALUES " +
               ",".join(f"({i},{i},{i*2})" for i in range(n_rows)))
    db.analyze()

    keys = [random.randrange(n_rows) for _ in range(probes)]

    idx_plan = db.explain(f"SELECT v FROM t WHERE id = {keys[0]}")
    seq_plan = db.explain(f"SELECT v FROM t WHERE shadow = {keys[0]}")

    t0 = time.perf_counter()
    for k in keys:
        list(db.execute(f"SELECT v FROM t WHERE id = {k}"))
    idx_t = time.perf_counter() - t0

    t0 = time.perf_counter()
    for k in keys:
        list(db.execute(f"SELECT v FROM t WHERE shadow = {k}"))
    seq_t = time.perf_counter() - t0
    db.close()
    return {
        "n_rows": n_rows, "probes": probes,
        "index_total_s": idx_t, "seq_total_s": seq_t,
        "index_us_per_query": idx_t / probes * 1e6,
        "seq_us_per_query": seq_t / probes * 1e6,
        "speedup": seq_t / idx_t,
        "index_plan": idx_plan.splitlines()[-1].strip(),
        "seq_plan": seq_plan.splitlines()[-1].strip(),
    }


# ---------------------------------------------------------------------------
# Experiment 3 — buffer pool hit rate
# ---------------------------------------------------------------------------
def run_buffer(n_rows=20000, capacity=64, passes=3):
    db = fresh("/tmp/minidb_bench_buf", "2PL")
    db.bufferpool.capacity = capacity
    db.execute("CREATE TABLE t (id INT PRIMARY KEY, v INT)")
    db.execute("INSERT INTO t VALUES " +
               ",".join(f"({i},{i})" for i in range(n_rows)))
    # reset counters after load
    db.bufferpool.hits = db.bufferpool.misses = db.bufferpool.evictions = 0
    for _ in range(passes):
        for _r in db.heaps["t"].scan():
            pass
    s = db.bufferpool.stats()
    pages = db.disk.num_pages("t")
    db.close()
    total = s["hits"] + s["misses"]
    return {
        "rows": n_rows, "pages": pages, "capacity": capacity, "passes": passes,
        "hits": s["hits"], "misses": s["misses"], "evictions": s["evictions"],
        "hit_rate": (s["hits"] / total) if total else 0.0,
    }


# ---------------------------------------------------------------------------
def main():
    lines = []

    def out(s=""):
        print(s)
        lines.append(s)

    out("# MiniDB Benchmark Results")
    out()
    out(f"_Generated by `benchmarks/bench.py` on Python "
        f"{sys.version.split()[0]}._")
    out()

    # Exp 1
    out("## 1. Concurrency — 2PL vs MVCC under read/write contention")
    out()
    out("Workload: 8 reader threads (point SELECT by primary key) and 2 writer "
        "threads (delete+insert on a hot row, holding it ~12 ms) over a "
        "3000-row table whose accesses concentrate on a 40-row hot set, for "
        "~2.5 s. Readers in 2PL must wait for a writer's exclusive lock on the "
        "same row; MVCC readers see a consistent snapshot and never block.")
    out()
    two = run_contention("2PL")
    mv = run_contention("MVCC")
    two_w = run_writers_only("2PL")
    mv_w = run_writers_only("MVCC")
    out("| Metric | 2PL | MVCC |")
    out("|---|---:|---:|")
    out(f"| Read throughput (txn/s) | {two['read_tps']:.0f} | {mv['read_tps']:.0f} |")
    out(f"| Avg read latency (ms) | {two['avg_read_latency_ms']:.3f} | {mv['avg_read_latency_ms']:.3f} |")
    out(f"| Reads completed | {two['reads_ok']} | {mv['reads_ok']} |")
    out(f"| Write throughput (txn/s, writers-only) | {two_w['write_tps']:.0f} | {mv_w['write_tps']:.0f} |")
    speedup = mv['read_tps'] / two['read_tps'] if two['read_tps'] else float('nan')
    out()
    out(f"**MVCC read throughput is {speedup:.2f}x that of 2PL**, with lower "
        f"read latency, because snapshot reads do not block on writers. This is "
        f"the core result for Extension Track B.")
    out()
    out("> _Caveat:_ write throughput is reported from a writers-only run. Under "
        "CPython's global interpreter lock, MVCC's lock-free readers busy-run "
        "and starve writer threads, while 2PL readers block (yielding the GIL); "
        "measuring writes without readers removes that interpreter artifact and "
        "isolates the engine's behaviour.")
    out()

    # Exp 2
    out("## 2. Access methods — B+ tree IndexScan vs SeqScan")
    out()
    am = run_access_methods()
    out(f"Workload: {am['probes']} point lookups over a {am['n_rows']}-row table. "
        f"The primary key is indexed; an identical un-indexed `shadow` column "
        f"forces a full scan.")
    out()
    out("| Access path | Plan | µs / query | Total (s) |")
    out("|---|---|---:|---:|")
    out(f"| IndexScan (PK) | `{am['index_plan']}` | {am['index_us_per_query']:.1f} | {am['index_total_s']:.3f} |")
    out(f"| SeqScan (shadow) | `{am['seq_plan']}` | {am['seq_us_per_query']:.1f} | {am['seq_total_s']:.3f} |")
    out()
    out(f"**The index is ~{am['speedup']:.0f}x faster** for selective point "
        f"lookups — O(log n) tree descent vs O(n) scan.")
    out()

    # Exp 3
    out("## 3. Buffer pool — hit rate vs capacity")
    out()
    rows = []
    for cap in (16, 64, 256):
        rows.append(run_buffer(capacity=cap))
    out(f"Workload: {rows[0]['passes']} sequential scans of a "
        f"{rows[0]['rows']}-row table ({rows[0]['pages']} pages) at three pool "
        f"sizes. LRU keeps recently-touched pages resident.")
    out()
    out("| Pool capacity (pages) | Hits | Misses | Evictions | Hit rate |")
    out("|---:|---:|---:|---:|---:|")
    for r in rows:
        out(f"| {r['capacity']} | {r['hits']} | {r['misses']} | "
            f"{r['evictions']} | {r['hit_rate']*100:.1f}% |")
    out()
    out("When the pool can hold the whole file the second and third passes are "
        "almost entirely hits; when it is far smaller than the file, sequential "
        "scanning evicts pages before they are reused and the hit rate collapses "
        "(classic scan/LRU behaviour).")
    out()

    with open(OUT, "w") as f:
        f.write("\n".join(lines) + "\n")
    print(f"\n[written {OUT}]")


if __name__ == "__main__":
    main()
